### PR TITLE
feat(HMS-1246): Use permission checks in statuser

### PR DESCRIPTION
**Pull request by avitova**
This adds permission logging in case of missing permissions. Caching should also be done for this. I feel like the source result should not be flagged as Unavailable or with Error in case of missing permission, as this is neither of these. Therefore, I thought adding the MissingPermission parameter might be suitable, but let me know, what you think.
For testing the missing permission in statuser, you could change iam_client hardcoded struct.

### CODE REVIEW

**Comment by lzap** [comment]
> I feel like the source result should not be flagged as Unavailable or with Error in case of missing permission, as this is neither of these.

That is correct, I think a warning in logs is enough. We should not see many of these, only old sources would experience this.

**Comment by lzap** [comment]
Test failure seems relevant (timeout)?

File: cmd/pbstatuser/main.go [hunk 1]
```diff
@@ -151,16 +151,24 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
...
 				ResourceType: "Application",
 			}
-			_, err = clients.GetEC2Client(ctx, &s.Authentication, "")
+			ec2Client, err := clients.GetEC2Client(ctx, &s.Authentication, "")
 			if err != nil {
 				sr.Status = kafka.StatusUnavailable
 				sr.Err = err
-				logger.Warn().Err(err).Msg("Could not get aws assumed client")
-				chSend <- sr
+				logger.Warn().Err(err).Msgf("Could not get aws assumed client %s", err)
```
Comment (lzap) [inline cmd/pbstatuser/main.go:158]: Please do not send error messages to messages because of Sentry.

https://github.com/RHEnVision/provisioning-backend/pull/488

File: cmd/pbstatuser/main.go [hunk 2]
```diff
@@ -151,16 +151,24 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
...
+				sr.Status = kafka.StatusAvailable
+			}
+			permissions, err := ec2Client.CheckPermission(ctx, &s.Authentication)
+			if err != nil {
+				sr.Err = err
+				logger.Warn().Err(err).Msgf("Could not check aws permissions %s", err)
+			}
+			sr.MissingPermissions = permissions
+			if len(permissions) != 0 {
+				logger.Warn().Err(err).Msgf("No sufficient permissions: %s", permissions)
```
Comment (lzap) [inline cmd/pbstatuser/main.go:169]: This does not seem correct. CheckPermissions either returns nil result and error, or something in the result. You reported the error above, here I would simply issue a warning with the list. Also can you add a field so we can easily find all records in Kibana, for example:

    logger.Warn().Bool("missing_permission", true).Msgf("No sufficient permissions: %s", permissions)

File: cmd/pbstatuser/main.go [hunk 3]
```diff
@@ -151,16 +151,24 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
...
+				logger.Warn().Err(err).Msgf("Could not get aws assumed client %s", err)
 			} else {
-				sr.Status = kafka.StatusAvaliable
-				chSend <- sr
+				sr.Status = kafka.StatusAvailable
+			}
+			permissions, err := ec2Client.CheckPermission(ctx, &s.Authentication)
+			if err != nil {
+				sr.Err = err
+				logger.Warn().Err(err).Msgf("Could not check aws permissions %s", err)
```
Comment (lzap) [inline cmd/pbstatuser/main.go:165]: Ditto.

File: internal/kafka/source_result.go [hunk 4]
```diff
@@ -24,6 +24,8 @@ type SourceResult struct {
 	Err error `json:"error"`
 
 	Identity identity.XRHID `json:"-"`
+
+	MissingPermissions []string `json:"missing_permissions"`
 }
```
Comment (lzap) [inline internal/kafka/source_result.go:29]: Sources do not currently support that, so I would use `-` json tag for now. Only into logging for now.

**Review by lzap — COMMENTED** [review]
Looks great, there is a failure can you take a look?

**Comment by lzap** [comment]
_For the record, @avitova might not get to working on this PR. This is not high-priority so let’s keep it opened._

**Comment by lzap** [comment]
Let me know when this is ready for re-review.

File: cmd/pbackend/statuser.go [hunk 5]
```diff
@@ -131,7 +131,7 @@ func checkSourceAvailabilityAzure(ctx context.Context) {
 				ResourceType: "Application",
 			}
 			// TODO: check if source is avavliable - WIP
```
Comment (lzap) [inline cmd/pbackend/statuser.go:133]: Can you drop this TODO? This patch actually implements the missing functionality.

File: cmd/pbackend/statuser.go [hunk 6]
```diff
@@ -153,16 +153,24 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
...
+				sr.Status = kafka.StatusAvailable
 			}
+			permissions, err := ec2Client.CheckPermission(ctx, &s.Authentication)
+			if err != nil {
+				sr.Err = err
+				logger.Warn().Err(err).Msg("AWS permission check failed")
+			}
+			sr.MissingPermissions = permissions
+			if permissions != nil {
+				logger.Warn().Bool("missing_permission", true).Msgf("No sufficient permissions: %s", permissions)
```
Comment (lzap) [inline cmd/pbackend/statuser.go:171]: Perhaps "insufficient" is better wording.

Also you can use `strings.Join` to join string slice by comma, but [this works too](https://go.dev/play/p/yVujHTNW-IL).

Finally, when `CheckPermission` returns `nil` it likely returns an error too, can you squash this into just a single warning message? You can stuff everything into one log record, me as someone who is debugging this issue would be interested in:

* source id (already present in the context logger)
* account number (could be added in the `processMessage` too as it is in the identity but perhaps out of scope of this PR)
* error message (via Err function)
* missing permissions

So basically this line contains it all, perhaps just delete the 167 and move this into the if statement above or something like that.

File: cmd/pbackend/statuser.go [hunk 7]
```diff
@@ -153,16 +153,24 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
...
 				logger.Warn().Err(err).Msg("Could not get aws assumed client")
-				chSend <- sr
 			} else {
-				sr.Status = kafka.StatusAvaliable
-				chSend <- sr
+				sr.Status = kafka.StatusAvailable
 			}
+			permissions, err := ec2Client.CheckPermission(ctx, &s.Authentication)
+			if err != nil {
+				sr.Err = err
```
Comment (lzap) [inline cmd/pbackend/statuser.go:166]: Here you should also set `sr.Status = kafka.StatusUnavailable` because missing permissions effectively means the source is not "all right" and the user should delete it and create it again.

File: cmd/pbackend/statuser.go (hunk 6 above)
Comment (avitova) [reply cmd/pbackend/statuser.go]: This is a very good point. The CheckPermission function returned only missing permissions when something was missing, and the error value was returned only when something else failed. We should return an error for missing permissions, so I changed it in this PR. I hope this now makes more sense.

**Comment by ezr-ondrej** [comment]
:tada: ... no idea what was the failure before, but now it works :smirk:

File: cmd/pbackend/statuser.go [hunk 8]
```diff
@@ -153,16 +152,22 @@ func checkSourceAvailabilityAWS(ctx context.Context) {
 				Identity:     s.Identity,
 				ResourceType: "Application",
 			}
-			_, err = clients.GetEC2Client(ctx, &s.Authentication, "")
+			ec2Client, err := clients.GetEC2Client(ctx, &s.Authentication, "")
 			if err != nil {
 				sr.Status = kafka.StatusUnavailable
 				sr.Err = err
 				logger.Warn().Err(err).Msg("Could not get aws assumed client")
```
Comment (ezr-ondrej) [inline cmd/pbackend/statuser.go:159]: Shouldn't we make sure, that when `GetEC2Client` errors out, we do not call CheckPermission on it? that might be nil pointer reference, right?

I'd just return here honestly, we would need to bump the metric here as well, but that's just three extra lines (write to channel, bump metric and return), not a big deal. WDYT?

Comment (ezr-ondrej) [reply cmd/pbackend/statuser.go:159]: I might be missing something here tho :)

Comment (avitova) [reply cmd/pbackend/statuser.go:159]: That's definitely true, we should return here! TY.

Comment (lzap) [reply cmd/pbackend/statuser.go:159]: Guys, please no `return` statements here, this is a channel processing loop!

One note: this is one of rare examples when `goto` statement which is [available in Go](https://go.dev/ref/spec#Goto_statements) could be reasonable, also a function could be good fit. I see some code being duplicated (creating the result, sending the result to).

Comment (ezr-ondrej) [reply cmd/pbackend/statuser.go:160]: I believe @avitova solved it pretty well :)
Good compromise! :+1:

**Review by ezr-ondrej — APPROVED** [review]
I'm good here, thanks! :orange_heart:

File: cmd/pbackend/statuser.go (hunk 8 above)
Comment (lzap) [reply cmd/pbackend/statuser.go:160]: Oh no, did I scared you, @ezr-ondrej ? :-)

You merge me a trigger and stored SQL procedure and the next day I debate loops :-D

**Review by lzap — APPROVED** [review]
One detail - the error length is unlimited and in case user actually forgot to add any permission this gets rendered to screen as something very long (covering whole screen or something).

In another PR, I suggest to limit the list to some reasonable length and in these cases add ` and more` to the error.
